monitoring/prometheus/exporter.go

```diff
@@ -0,0 +1,194 @@
+
+var errKeyNotFound = errors.New("key not found")
+
+// NewExporter returns an initialized exporter.
+func NewExporter(server, namespace string, timeout time.Duration) *Exporter {
+	return &Exporter{
+		address:   server,
+		timeout:   timeout,
+		namespace: namespace,
+		up: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "up"),
+			"If tinode server is reachable.",
+			nil,
+			nil,
+		),
+		uptime: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "uptime_seconds"),
+			"Number of seconds since the server started.",
+			nil,
+			nil,
+		),
+		version: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "version"),
+			"The version of this tinode server.",
+			[]string{"version"},
+			nil,
+		),
+		topicsLive: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "topics_live_count"),
+			"Number of currenly active topics.",
+			nil,
+			nil,
+		),
+		topicsTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "topics_total"),
+			"Total number of topics used during server lifetime.",
+			nil,
+			nil,
+		),
+		sessionsLive: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "sessions_live_count"),
+			"Number of currenly active sessions.",
+			nil,
+			nil,
+		),
+		sessionsTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "sessions_total"),
+			"Total number of sessions since server start.",
+			nil,
+			nil,
+		),
+		malloced: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "malloced_bytes"),
+			"Number of bytes of memory allocated and in use.",
+			nil,
+			nil,
+		),
+	}
+}
+
+// Describe describes all the metrics exported by the memcached exporter. It
+// implements prometheus.Collector.
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	ch <- e.up
+	ch <- e.uptime
+	ch <- e.version
+	ch <- e.topicsLive
+	ch <- e.topicsTotal
+	ch <- e.sessionsLive
+	ch <- e.sessionsTotal
+	ch <- e.malloced
+}
+
+// Collect fetches statistics from the configured Tinode instance, and
+// delivers them as Prometheus metrics. It implements prometheus.Collector.
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	resp, err := http.Get(e.address)
+	if err != nil {
+		ch <- prometheus.MustNewConstMetric(e.up, prometheus.GaugeValue, 0)
+		log.Println("Failed to connect to server", err)
+		return
+	}
+	defer resp.Body.Close()
+
+	up := float64(1)
+
+	var stats map[string]interface{}
+	err = json.NewDecoder(resp.Body).Decode(&stats)
+	if err != nil {
+		log.Println("Failed to fetch or parse response", err)
+		up = 0
+	} else {
+		if err := e.parseStats(ch, stats); err != nil {
+			up = 0
+		}
+	}
+
+	ch <- prometheus.MustNewConstMetric(e.up, prometheus.GaugeValue, up)
+}
+
+func (e *Exporter) parseStats(ch chan<- prometheus.Metric, stats map[string]interface{}) error {
+
+	err := firstError(
+		e.parseAndUpdate(ch, e.version, prometheus.GaugeValue, stats, "Version"),
+		e.parseAndUpdate(ch, e.uptime, prometheus.CounterValue, stats, "uptime"),
```

Comment:
I don't see these metrics in the expvar. Where are these coming from?

Reply:
`Version` is added as a part of this PR: https://github.com/tinode/chat/pull/277/files#diff-d3d4b092ce8121dc129c8b29e3545206R549

`uptime` was left for reference. I've removed it now.